### ISSUE TITLE
Ignore non-significant whitespace in fixed format parser

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#779]
SciPy uses (only at one place) syntax for an `if` statement that I have not seen before. The following patch fixes it:
```diff
--- a/scipy/special/specfun/specfun.f
+++ b/scipy/special/specfun/specfun.f
@@ -2761,8 +2761,11 @@ C
               F2=F1
 50            F1=F
            CS=0.0D0
-           IF (DABS(BJV0).GT.DABS(BJV1)) CS=BJV0/F
-           ELSE CS=BJV1/F2
+           IF (DABS(BJV0).GT.DABS(BJV1)) then
+               CS=BJV0/F
+           ELSE
+               CS=BJV1/F2
+           end if
            DO 55 K=0,N
 55            VL(K)=CS*VL(K)
         ENDIF
```
We need to investigate as this might be a bug in the source code, as the `ELSE` possibly attaches to a previous if; and the only extension is to use a statement at the same line as `else`.

**Comment by konradha** [comment]
I'm pretty sure it's just the indentation that's off. The entire block is

```fortran
       IF (N.GE.2.AND.N.LE.INT(0.9*X)) THEN
           F0=BJV0
           F1=BJV1
           DO 45 K=2,N
              F=2.0D0*(K+V0-1.0D0)/X*F1-F0
              F0=F1
              F1=F
45            VL(K)=F
        ELSE IF (N.GE.2) THEN
           M=MSTA1(X,200)
           IF (M.LT.N) THEN
              N=M
           ELSE
              M=MSTA2(X,N,15)
           ENDIF
           F=0.0D0
           F2=0.0D0
           F1=1.0D-100
           DO 50 K=M,0,-1
              F=2.0D0*(V0+K+1.0D0)/X*F1-F2
              IF (K.LE.N) VL(K)=F
              F2=F1
50            F1=F
           CS=0.0D0
           IF (DABS(BJV0).GT.DABS(BJV1)) CS=BJV0/F
           ELSE CS=BJV1/F2
           DO 55 K=0,N
55            VL(K)=CS*VL(K)
        ENDIF
```

The entire block starts with `IF () THEN` then `ELSE IF () THEN` then `ELSE` -- the single-line `IF` is just confusing but I think what you're proposing is not the actual behavior of the code. I will explore this a little later but I don't think you can chain single-line `IF` statements with other conditionals.

**Comment by awvwgk** [comment]
This is a bug in scipy, I think Ondrej's proposed fix is the intended implementation, while the actual implementation uses

```f90
cs = 0.0d0
if (abs(bjv0) > abs(bjv1)) cs = bjv0 / f
elsecs = bjv1 / f2
```

Note that the “else” branch will never be reached because it is an assignment to the implicitly typed variable `elsecs`, which remains unused in the later code.

**Comment by konradha** [comment]
Thanks, that makes sense.

**Comment by awvwgk** [comment]
Reported at https://github.com/scipy/scipy/issues/17104

**Comment by awvwgk** [comment]
While the code in scipy is incorrect, the required change for LFortran would be to identify the implicitly typed variable *elsecs* and perform the assignment. A diagnostic in the fixed format parser when ignoring non-significant whitespace might be useful, in most cases this is likely to be a programmer error.

**Comment by certik** [comment]
Ha, amazing!!

@jacobwilliams, @fortranfan you will both love this, another example of the default Fortran behavior causing a bug, and there being no issue with backwards compatibility (for SciPy) due to the fact that LFortran currently can't handle a variable `elsecs` in fixed form. For the background: in the last couple days I got our fixed-form parser parse all of SciPy, and I noted a few things that still need fixing, such as the `elsecs` issue, you can see it in my comment here: https://github.com/lfortran/lfortran/blob/41532ef8f936b9b7f86d07c3f7a402681eb67f18/src/lfortran/parser/fixedform_tokenizer.cpp#L1181, I am aware of it, but wanted to first parse SciPy, then worry about these corner cases. Due to this, LFortran currently refuses to compile `elsecs`, and it turns out this was an actual bug in SciPy.... 

Obviously we'll "fix" it in LFortran, as we want to parse code according to the standard.

However, probably adding some optional warnings for such usage might be very helpful to catch bugs. Of course, I don't know if it is worth investing more into fixed-form, as I really want to focus my efforts into free-form, where this particular issue does not happen. The only advantage would be to catch bugs in existing fixed-form code, such as the above bug.

Let me know your thoughts.

**Comment by jacobwilliams** [comment]
Ug, this is the perfect storm of Fortran 77 nonsense!

**Comment by Thirumalai-Shaktivel** [comment]
That's great news! I'm glad that our LFortran parser has become very strong.

**Comment by FortranFan** [comment]
> Ha, amazing!!
> 
> .. @FortranFan you will both love this, another example of the default Fortran behavior causing a bug, and there being no issue with backwards compatibility (for SciPy) due to the fact that LFortran currently can't handle a variable `elsecs` in fixed form. For the background: in the last couple days I got our fixed-form parser parse all of SciPy, and I noted a few things that still need fixing, such as the `elsecs` issue, you can see it in my comment here:
> ..
> Let me know your thoughts.


Thanks for sharing @certik .  Interesting situation and disappointing also, given certain intransigent positions when it comes to moving forward to overcome the limitations with legacy semantics in `Fortran`.

For the immediate situation with `specfun.f` in SciPy, both you and @awvwgk have converged on the right approach.

For the longer term, I wonder whether we can review the special functions of interest to SciPy and cross-check against the current Fortran standard (it already includes support for various aspects including Bessel functions) and also the Fortran `stdlib` efforts.

Ultimately it will be good if the base Fortran language includes as much support as reasonable toward special functions followed by robust coverage in the Fortran `stdlib` to arrive at a point where SciPy may feel compelled to do away with the legacy FORTRAN implementation in `specfun.f`, notwithstanding their own backward compatibility challenges with the earlier, poorly written code.

By the way, do you have any info on the genesis of `specfun.f`?  How did it get authored, under whose influence, and when?  It appears to have been written in not too distant past i.e., after the Fortran 90/95 standard revision and in which case it is disappointing it did not use any of the good coding practices made possible with Fortran 90/95.

**Comment by certik** [comment]
@FortranFan thanks for the comment. I think the ultimate goal is to convert the SciPy code to free form and modern Fortran, ideally automatically (currently the fixed-form tokenizer drops all comments, but if we preserve them, our AST can contain comments and the formatter prints them, so that should work). And then improving their code and algorithms, in Fortran. So yes, the `specfun.f` can and should be greatly modernized. :)

**Comment by certik** [comment]
I opened up an issue for this: https://github.com/lfortran/lfortran/issues/815. As mentioned there, we currently drop all spaces, so we do not have the information available to issue such a warning.

If it was easy to do, I would do it right away. Given our current fixed-form tokenizer, it seems it would require a significant redesign, which I don't think is a high priority right now. I think the highest priorities (at least for me) are:

* Get fixed-form tokenizer to parse SciPy (almost there)
* Get semantics and LLVM working for SciPy
* Get semantics and LLVM working for stdlib and fpm
* Work on good optimizations (ASR->ASR)
* Get fixed-form tokenizer to parse other codes, as the need arises or people request it / report bugs; we have designed the tokenizer to be easily extensible to parse (valid) code.

In general I want to focus on getting LFortran from alpha to beta, and to wide usage. We definitely want the best warnings in free-form code, and I think we have all the information that we need for that. I think for example the "implicit save" warnings are easy to do and we'll do it.

The warnings specific to fixed-form, especially if it would require to completely rewrite it again, is low in my priority list. Of course, if somebody is interested in doing the work, then let's do it. But I would recommend to focus on the free-form and modern features as the community, and just do enough solid work for old features so that we can use them.
